dix/req.c

```diff
@@ -1625,24 +1628,31 @@ static void dix_rop_completed(struct m0_sm_group *grp, struct m0_sm_ast *ast)
 	struct m0_dix_rop_ctx *rop_del_phase2 = NULL;
 	bool                   del_phase2 = false;
 	struct m0_dix_cas_rop *cas_rop;
+	int                    successful_ops = 0;
 
 	(void)grp;
 	if (req->dr_type == DIX_NEXT)
 		m0_dix_next_result_prepare(req);
 	else {
+		m0_tl_for (cas_rop, &rop->dg_cas_reqs, cas_rop) {
+			if (cas_rop->crp_creq.ccr_sm.sm_rc == 0) {
+				successful_ops++;
+			}
+		} m0_tl_endfor;
+
 		/*
-		 * Consider DIX request to be successful if there is at least
-		 * one successful CAS request.
+		 * If enough operations succeeded to satisfy the quorum requirement,
+		 * ignore any operations that failed. If we don't have a quorum,
+		 * count all operations when deciding overall success (ensures the
+		 * overall operation is considered a failure).
 		 */
-		if (m0_tl_forall(cas_rop, cas_rop,
-				 &rop->dg_cas_reqs,
-				 cas_rop->crp_creq.ccr_sm.sm_rc != 0))
-			    dix_cas_rop_rc_update(cas_rop_tlist_tail(
-						  &rop->dg_cas_reqs), 0);
 
+		M0_ASSERT(req->dr_quorum > 0);
```

Comment:
This should be checked much earlier. The best way is to introduce `bool m0_dix_req_invariant(const struct m0_dix_req *req)` that will check this (and maybe some other request consistency conditions) and call in on entry and exit from request manipulating functions (see existing invariants). But at least, quorum > 0 should be checked when the operation is initialised.

Reply:
I had checked back in ` m0_idx_op`, but you're right, this should be checked during struct initialization. Added more checks